src/main/java/io/cdap/plugin/gcp/firestore/source/QueryDocumentSnapshotToRecordTransformer.java

```diff
@@ -0,0 +1,223 @@
+    this.schema = schema;
+    this.includeDocumentId = includeDocumentId;
+    this.idAlias = idAlias;
+  }
+
+  /**
+   * Transforms given {@link QueryDocumentSnapshot} to {@link StructuredRecord}.
+   *
+   * @param queryDocumentSnapshot queryDocumentSnapshot object to be transformed.
+   * @return {@link StructuredRecord} that corresponds to the given {@link QueryDocumentSnapshot}.
+   */
+  public StructuredRecord transform(QueryDocumentSnapshot queryDocumentSnapshot) {
+    return convertRecord(null, queryDocumentSnapshot, schema);
+  }
+
+  private StructuredRecord convertRecord(@Nullable String fieldName, QueryDocumentSnapshot object, Schema schema) {
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+    List<Schema.Field> fields = Objects.requireNonNull(schema.getFields(), "Schema fields cannot be empty");
+    for (Schema.Field field : fields) {
+      if (includeDocumentId && idAlias.equals(field.getName())) {
+        builder.set(field.getName(), object.getId());
+      } else {
+        // Use full field name for nested records to construct meaningful errors messages.
+        // Full field names will be in the following format: 'record_field_name.nested_record_field_name'
+        String fullFieldName = fieldName != null ? String.format("%s.%s", fieldName, field.getName()) :
+          field.getName();
+
+        Schema nonNullableSchema = field.getSchema().isNullable() ? field.getSchema().getNonNullable()
+          : field.getSchema();
+        Object value = extractValue(fullFieldName, object.get(field.getName()), nonNullableSchema);
+        if (!field.getSchema().isNullable() && value == null) {
+          builder.set(field.getName(), "");
+        } else {
+          builder.set(field.getName(), value);
+        }
+      }
+    }
+    return builder.build();
+  }
+
+  private Object extractValue(String fieldName, Object object, Schema schema) {
+    //if (object == null || object instanceof BsonUndefined) {
+    if (object == null) {
+      return null;
+    }
+
+    Schema.LogicalType fieldLogicalType = schema.getLogicalType();
+    if (fieldLogicalType != null) {
+      switch (fieldLogicalType) {
+        case TIMESTAMP_MILLIS:
+          ensureTypeValid(fieldName, object, Date.class);
+          Instant instant = ((Date) object).toInstant();
+          long millis = TimeUnit.SECONDS.toMillis(instant.getEpochSecond());
+          return Math.addExact(millis, TimeUnit.NANOSECONDS.toMillis(instant.getNano()));
+        case TIMESTAMP_MICROS:
+          ensureTypeValid(fieldName, object, Date.class);
+          Instant dateInstant = ((Date) object).toInstant();
+          long micros = TimeUnit.SECONDS.toMicros(dateInstant.getEpochSecond());
+          return Math.addExact(micros, TimeUnit.NANOSECONDS.toMicros(dateInstant.getNano()));
+          /*
+        case DECIMAL:
+          ensureTypeValid(fieldName, object, Decimal128.class);
```

Comment:
why do we comment out this, how is the decimal type handled now?

Reply:
Done